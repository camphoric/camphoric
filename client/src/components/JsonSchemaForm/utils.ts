import getFromPath from 'lodash/get';

export const getSchemaValue = (props: object, name: string): string | undefined => 
  getFromPath(props, ['schema', name])
      || getFromPath(props, ['uiSchema', `ui:${name}`])
      || getFromPath(props, name);

export const getSchemaItemsValue = (props: object, name: string): string | undefined => 
  getFromPath(props, ['schema', 'items', name])
      || getFromPath(props, ['uiSchema', 'items', `ui:${name}`])
      || getFromPath(props, name);

