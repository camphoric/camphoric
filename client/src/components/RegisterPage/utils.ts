import jsonLogic from "json-logic-js";
import { JSONSchema6 } from "json-schema";

import { RegistrationConfig, FormData } from "./RegisterPage";

type PricingData = {
  event: { [key: string]: any };
  registration: FormData;
  pricing: { [key: string]: any };
  [key: string]: any;
};

type PricingResults = {
  [key: string]: any;
  campers: Array<{ [key: string]: any }>;
  total: number;
};

export function calculatePrice(config: RegistrationConfig, formData: FormData) {
  // calculation should be done in whole dollars for sake of
  // avoiding funky issues with floats. If sub-dollar amounts
  // are necessary, we should switch this to cents.

  const { event, pricingLogic, pricing } = config;

  const results: PricingResults = {
    campers: [],
    total: 0,
  };

  const data: PricingData = {
    event,
    registration: formData,
    pricing,
  };

  pricingLogic.registration.forEach(component => {
    const varName = component.var;
    const value = jsonLogic.apply(component.exp, data);
    results[varName] = isNaN(value) ? 0 : value;
    data[varName] = value;
  });

  formData.campers.forEach(camper => {
    data.camper = camper;
    const camperResults: { [key: string]: any } = {};
    pricingLogic.camper.forEach(component => {
      const varName = component.var;
      const value = jsonLogic.apply(component.exp, data);
      camperResults[varName] = value;
      if (typeof value === 'number' || typeof value === 'boolean') {
        const subtotal = (results[varName] || 0) + value;
        results[varName] = subtotal;
        data[varName] = value;
      }
    });
    results.campers.push(camperResults);
  });

  return results;
}
