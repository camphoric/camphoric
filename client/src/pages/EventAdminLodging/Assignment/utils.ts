import { LodgingLookup } from 'hooks/api';

export function getDaysArray(start: string, end: string): Array<Date> {
  let arr = [];

  for(
    let dt = new Date(start);
    dt <= new Date(end);
    dt.setDate(dt.getDate() + 1)
  ){
    arr.push(new Date(dt));
  }

  return arr;
};

export function getLeafNodes(
  lodgingTree: AugmentedLodging,
  leaves: LodgingLookup = {},
  titles: Array<string> = []
): LodgingLookup {
  if (lodgingTree.isLeaf) {
    leaves[[
      ...titles,
      lodgingTree.name,
    ].join('->')] = lodgingTree;
    return leaves;
  }

  let newTitles = titles;
  if (Object.values(leaves).length) {
    newTitles = [...titles, lodgingTree.name || '?'];
  }

  lodgingTree.children.forEach(
    (l) => getLeafNodes(l, leaves, newTitles)
  );

  return leaves;
};

export type LodgingPair = [a: string, b: AugmentedLodging];

export const sortLodgingNames = (a: LodgingPair, b: LodgingPair) => {
  const nameA = a[0];
  const nameB = b[0];
  if (nameA < nameB) {
    return -1;
  }
  if (nameA > nameB) {
    return 1;
  }

  // names must be equal
  return 0;
};

