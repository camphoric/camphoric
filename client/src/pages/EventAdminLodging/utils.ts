import { LodgingLookup } from 'hooks/api';
import flattenDeep from 'lodash/flattenDeep';

export function getDaysArray(event?: ApiEvent): Array<Date> {
  let arr: Array<Date> = [];

  if (!event) return arr;
  if (!event.start || !event.end) return arr;

  for(
    let dt = new Date(event.start);
    dt <= new Date(event.end);
    dt.setDate(dt.getDate() + 1)
  ){
    arr.push(new Date(dt));
  }

  return arr;
};

export function lodgingIsAncestorOf(
  leaf: AugmentedLodging,
  targetId: number | string,
  lookup: LodgingLookup,
) {
  const findInParent = (id?: string | number): boolean => {
    if (!id) return false;

    const node = lookup[id];

    if (node.id.toString() === targetId.toString()) return true;

    return findInParent(node.parent);
  };

  return findInParent(leaf.id);
}

export function getNonLeafNodes(
  lodgingTree: AugmentedLodging,
  nonLeaves: Array<AugmentedLodging> = [],
): Array<AugmentedLodging> {
  if (lodgingTree.isLeaf) return nonLeaves;

  nonLeaves.push(lodgingTree);

  lodgingTree.children.forEach(
    (l) => getNonLeafNodes(l, nonLeaves)
  );

  return nonLeaves.slice(1);
};


export function getLeafNodes(
  lodgingTree: AugmentedLodging,
  leaves: LodgingLookup = {},
  titles: Array<string> = []
): LodgingLookup {
  if (lodgingTree.isLeaf) {
    leaves[[
      ...titles,
      lodgingTree.name,
    ].join('->')] = lodgingTree;
    return leaves;
  }

  let newTitles = titles;
  if (Object.values(leaves).length) {
    newTitles = [...titles, lodgingTree.name || '?'];
  }

  lodgingTree.children.forEach(
    (l) => getLeafNodes(l, leaves, newTitles)
  );

  return leaves;
};

export type LodgingPair = [a: string, b: AugmentedLodging];

export const sortLodgingNames = (a: LodgingPair, b: LodgingPair) => {
  const nameA = a[0];
  const nameB = b[0];
  if (nameA < nameB) {
    return -1;
  }
  if (nameA > nameB) {
    return 1;
  }

  // names must be equal
  return 0;
};

export const getAllParentClasses = (target: HTMLDivElement | null): Array<string> => {
  if (!target) return [];

  return flattenDeep([
    ...Array.from(target.classList || []),
    ...getAllParentClasses(target.parentNode as HTMLDivElement),
  ])
};

