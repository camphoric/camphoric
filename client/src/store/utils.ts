import { CreateApiOptions } from '@reduxjs/toolkit/query/react';

type BuilderFn = Parameters<
  CreateApiOptions<any, any, any, any>['endpoints']
>[0];

export const factory = <ApiObjectName extends string, Builder extends BuilderFn>(builder: Builder) => {
  const getCreator = <R>(apiObjectName: ApiObjectName) =>
    builder.query<[R], Object | void>({
      query: (params: { [a: string]: string }) => {
        const path = `${apiObjectName.toLowerCase()}s/`;

        if (!params) return path;

        const searchParams = new URLSearchParams(Object.entries(params));
        searchParams.sort();
        const queryString = searchParams.toString();
        return queryString ? `${path}?${queryString}` : path;
      },
      providesTags: [apiObjectName],
    });

  const getByIdCreator = <R>(apiObjectName: ApiObjectName) =>
    builder.query<R, Scalar>({
      query: (id) => `${apiObjectName.toLowerCase()}s/${id}/`,
      providesTags: (result, error, id) => [{ type: apiObjectName, id }],
    });

  const updateCreator = <R extends { id: Scalar }>(
    apiObjectName: ApiObjectName,
    invalidatesTags?: Array<ApiObjectName>,
  ) =>
    builder.mutation<R, Partial<R> & Pick<R, 'id'>>({
      // note: an optional `queryFn` may be used in place of `query`
      query: ({ id, ...patch }) => ({
        url: `${apiObjectName.toLowerCase()}s/${id}/`,
        method: 'PATCH',
        body: patch,
      }),
      invalidatesTags: invalidatesTags || [apiObjectName],
    });

  const createCreator = <R extends {}>(
    apiObjectName: ApiObjectName,
    invalidatesTags?: Array<ApiObjectName>,
  ) =>
    builder.mutation<
      R,
      Omit<R, 'id' | 'created_at' | 'updated_at' | 'deleted_at'>
    >({
      // note: an optional `queryFn` may be used in place of `query`
      query: (body) => ({
        url: `${apiObjectName.toLowerCase()}s/`,
        method: 'POST',
        body,
      }),
      invalidatesTags: invalidatesTags || [apiObjectName],
    });

  const deleteCreator = <R extends { id: Scalar }>(
    apiObjectName: ApiObjectName,
    invalidatesTags?: Array<ApiObjectName>,
  ) =>
    builder.mutation<R, { id: Scalar }>({
      // note: an optional `queryFn` may be used in place of `query`
      query: ({ id }) => ({
        url: `${apiObjectName.toLowerCase()}s/${id}/`,
        method: 'DELETE',
      }),
      invalidatesTags: invalidatesTags || [apiObjectName],
    });

  return {
    getCreator,
    getByIdCreator,
    updateCreator,
    createCreator,
    deleteCreator,
  };
}
