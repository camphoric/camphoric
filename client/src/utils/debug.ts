/**
 * Debug functions
 *
 * These functions are used to console.log when in development only.
 */

// @ts-ignore
export default function debug(...args) {
  if (process.env.NODE_ENV === 'development') {
    console.log(...args);
  }
}

