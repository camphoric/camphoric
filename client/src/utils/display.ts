/**
 * These functions are used for quickly displaying info that a user might need
 */
import { LodgingLookup } from 'hooks/api';
import getFromPath from 'lodash/get';

export function getCamperDisplayId(camper: ApiCamper) {
  // gather possible identifiers
  const lastName = [
    getFromPath(camper, 'attributes.lastname'),
    getFromPath(camper, 'attributes.last_name'),
  ].find(Boolean);
  const firstName = [
    getFromPath(camper, 'attributes.firstname'),
    getFromPath(camper, 'attributes.first_name'),
  ].find(Boolean);
  const fullName = [
    getFromPath(camper, 'attributes.name'),
    getFromPath(camper, 'attributes.fullname'),
  ].find(Boolean);
  const email = [
    getFromPath(camper, 'attributes.mail'),
    getFromPath(camper, 'attributes.email'),
  ].find(Boolean);

  // try a series of possible identifiers and return the first one that works
  return [
    [firstName, lastName].join(' '),
    fullName,
    email,
    lastName,
    firstName,
    camper.id,
  ].find(Boolean);
}

export function lodgingPathDisplay(
  lodgingLookup: LodgingLookup,
  lodgingId?: string | number,
  parts: Array<string> = []
): string {
  if (!lodgingId) return '';

  const node = lodgingLookup[lodgingId];

  if (!node.parent) return parts.join('â†’');

  return lodgingPathDisplay(lodgingLookup, node.parent, [ node.name, ...parts ]);
};

const ordinalsLessThan10 = [
  'Zeroth',
  'First',
  'Second',
  'Third',
  'Fourth',
  'Fifth',
  'Sixth',
  'Seventh',
  'Eighth',
  'Ninth',
];

export function ordinal(i: number) {
  if (i < 10 && i > 0) return ordinalsLessThan10[i];

  const j = i % 10;
  const k = i % 100;

  if (j === 1 && k !== 11) return `${i}st`;
  if (j === 2 && k !== 12) return `${i}nd`;
  if (j === 3 && k !== 13) return `${i}rd`;

  return `${i}th`;
}

export function getAllStrings(arg: any): Array<string> {
  if (typeof arg === 'string') {
    return [arg];
  }

  // handle wrong types and null
  if (typeof arg !== 'object' || !arg) {
    return [];
  }

  return Object.keys(arg).reduce((acc: Array<string>, key: string) => {
    return [...acc, ...getAllStrings(arg[key])];
  }, []);
};
