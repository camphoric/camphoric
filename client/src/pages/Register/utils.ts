import React from 'react';
import { useHistory } from 'react-router-dom';
import api from 'store/register/api';
import debounce from 'lodash/debounce';
import debug from 'utils/debug';
import { calculatePrice }from 'components/JsonSchemaForm';

import {
  setRegFormData,
  setTotals,
  resetAll,
  useAppDispatch,
  setUpdating,
} from 'store/register/store';

const allSlashes = /\//g;
export const useNavigateToRegPage = () => {
  const history = useHistory();

  return (to: string) => {
    const pathParts = history.location.pathname.split('/').slice(0, -1);
    pathParts.push(to.replace(allSlashes, ''))

    history.push({
      ...history.location,
      pathname: pathParts.join('/'),
    });
  }
}

const getLocalStorageKey = (config: ApiRegister) => {
  const name = config.dataSchema.title || 'formData';
  const date = config.event.start;

  return `${name}, ${date.year}-${date.month}-${date.day}`;
}

export const useLocalStorageFormData = () => {
  const registrationApi = api.useGetRegistrationQuery();
  const [rehydrated, setRehydrated] = React.useState(false);
  const dispatch = useAppDispatch();

  const config = registrationApi.data;
  let key: string;

  if (config) {
    key = getLocalStorageKey(config);
  }

  return {
    // Save to both local storage, redux, and recalc price
    saveFormData: debounce(async (formData) => {
      if (!key) return;
      if (!config) throw new Error('registrationApi.data is not defined');

      debug('saving form data');
      await dispatch(() => setUpdating(true));
      try {
        dispatch(setRegFormData(formData));
        dispatch(
          setTotals(calculatePrice(config, formData))
        );

        localStorage.setItem(key, JSON.stringify(formData));
      } catch (e) {
        console.error(e);
      } finally {
        dispatch(setUpdating(false));
      }
    }, 600, {leading:false, trailing:true}),

    // status of rehydration
    rehydrated,

    // restore from local storage, save to redux, and recalc price
    rehydrateFormData: () => {
      if (!key || rehydrated) return;
      if (!config) throw new Error('registrationApi.data is not defined');

      let formData;
      try {
        const saved = localStorage.getItem(key);

        if (!saved) {
          debug('no local storage form data');
          setRehydrated(true);

          return;
        }

        formData = JSON.parse(saved);

        dispatch(setRegFormData(formData));
        dispatch(
          setTotals(calculatePrice(config, formData))
        );

        setRehydrated(true);
      } catch (e) {
        console.error(e);
      }

      return formData;
    },

    // at the end of process, reset all data
    resetFormData: () => {
      if (!key) return;

      try {
        localStorage.removeItem(key);
        dispatch(resetAll());
        setRehydrated(false);
      } catch (e) {
        console.error(e);
      }
    },
  };
}

