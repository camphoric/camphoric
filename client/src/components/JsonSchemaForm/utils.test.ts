import { describe, it, expect } from 'vitest'
import * as utils from './utils';

describe('JsonSchemaForm utils', () => {
  describe('adminUISchemea', () => {
    const sampleUISchema = {
      meals: {
        meal_plan: {
          'ui:placeholder': 'Choose an option',
          'ui:enumDisabled': [
            'All Meals',
            'Just Dinners',
            'All Meals - 1st half',
            'All Meals - 2nd half'
          ],
        },
      },
    };
    it('should remove all enumDisabled keys', () => {
      const result = utils.adminUISchemea(sampleUISchema);
      expect(result).toMatchSnapshot();
    });
  });
});
