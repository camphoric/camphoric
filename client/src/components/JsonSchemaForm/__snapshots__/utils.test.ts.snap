// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`JsonSchemaForm utils > adminUISchemea > should remove all enumDisabled keys 1`] = `
{
  "meals": {
    "meal_plan": {
      "ui:placeholder": "Choose an option",
    },
  },
}
`;
